From 5776b0140e56f3a9bba32f085fd13272fa2602f8 Mon Sep 17 00:00:00 2001
From: Li Nan <linan122@huawei.com>
Date: Tue, 4 Apr 2023 10:05:21 +0800
Subject: [PATCH 0068/1646] [Huawei] md: limit the number of io for raid1/10
 plug

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188628

--------------------------------

There is no limit to the number of io for raid10 plug, whitch may result
in excessive memory usage and potential softlockup when a large number of
io are submitted at once. Limit it and flush when the limit is exceeded.

Fixes: 57c67df48866 ("md/raid10: submit IO from originating thread instead of md thread.")
Signed-off-by: Li Nan <linan122@huawei.com>
---
 drivers/md/md.c     | 7 +++++++
 drivers/md/md.h     | 2 ++
 drivers/md/raid1.c  | 2 ++
 drivers/md/raid10.c | 2 ++
 4 files changed, 13 insertions(+)

diff --git a/drivers/md/md.c b/drivers/md/md.c
index ddc71decb142..306b294ba0b1 100644
--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -8739,6 +8739,13 @@ void md_allow_write(struct mddev *mddev)
 }
 EXPORT_SYMBOL_GPL(md_allow_write);
 
+void md_flush_plug_cb(struct blk_plug_cb *cb)
+{
+	list_del(&cb->list);
+	cb->callback(cb, false);
+}
+EXPORT_SYMBOL_GPL(md_flush_plug_cb);
+
 #define SYNC_MARKS	10
 #define	SYNC_MARK_STEP	(3*HZ)
 #define UPDATE_FREQUENCY (5*60*HZ)
diff --git a/drivers/md/md.h b/drivers/md/md.h
index 0954b9fe4cf6..e01204217aa1 100644
--- a/drivers/md/md.h
+++ b/drivers/md/md.h
@@ -21,6 +21,7 @@
 #include "md-cluster.h"
 
 #define MaxSector (~(sector_t)0)
+#define MAX_PLUG_COUNT 32
 
 /*
  * These flags should really be called "NO_RETRY" rather than
@@ -740,6 +741,7 @@ extern int mddev_init_writes_pending(struct mddev *mddev);
 extern bool md_write_start(struct mddev *mddev, struct bio *bi);
 extern void md_write_inc(struct mddev *mddev, struct bio *bi);
 extern void md_write_end(struct mddev *mddev);
+extern void md_flush_plug_cb(struct blk_plug_cb *cb);
 extern void md_done_sync(struct mddev *mddev, int blocks, int ok);
 extern void md_error(struct mddev *mddev, struct md_rdev *rdev);
 extern void md_finish_reshape(struct mddev *mddev);
diff --git a/drivers/md/raid1.c b/drivers/md/raid1.c
index 00915e6ec410..72512a015bb5 100644
--- a/drivers/md/raid1.c
+++ b/drivers/md/raid1.c
@@ -1561,6 +1561,8 @@ static void raid1_write_request(struct mddev *mddev, struct bio *bio,
 		if (plug) {
 			bio_list_add(&plug->pending, mbio);
 			plug->pending_cnt++;
+			if (plug->pending_cnt / MAX_PLUG_COUNT >= disks)
+				md_flush_plug_cb(cb);
 		} else {
 			spin_lock_irqsave(&conf->device_lock, flags);
 			bio_list_add(&conf->pending_bio_list, mbio);
diff --git a/drivers/md/raid10.c b/drivers/md/raid10.c
index 2411e92db237..1a718f63624a 100644
--- a/drivers/md/raid10.c
+++ b/drivers/md/raid10.c
@@ -1291,6 +1291,8 @@ static void raid10_write_one_disk(struct mddev *mddev, struct r10bio *r10_bio,
 	if (plug) {
 		bio_list_add(&plug->pending, mbio);
 		plug->pending_cnt++;
+		if (plug->pending_cnt / MAX_PLUG_COUNT >= conf->copies)
+			md_flush_plug_cb(cb);
 	} else {
 		spin_lock_irqsave(&conf->device_lock, flags);
 		bio_list_add(&conf->pending_bio_list, mbio);
-- 
2.17.1

